xds/src/main/java/io/grpc/xds/internal/sds/ServerWrapperForXds.java

```diff
@@ -147,38 +148,45 @@ public Server start() throws IOException {
         new XdsClientWrapperForServerSds.ServerWatcher() {
           @Override
           public void onError(Throwable throwable, boolean isAbsent) {
-            if (currentServingState == ServingState.STARTING) {
-              // during start
-              if (isPermanentErrorFromXds(throwable)) {
-                currentServingState = ServingState.SHUTDOWN;
-                future.set(throwable);
+            synchronized (ServerWrapperForXds.this) {
+              if (currentServingState == ServingState.SHUTDOWN) {
                 return;
-              }
-              xdsServingStatusListener.onNotServing(throwable);
-            } else {
-              // after start
-              if (isAbsent) {
+              } else if (currentServingState == ServingState.STARTING) {
+                // during start
+                if (isPermanentErrorFromXds(throwable)) {
+                  currentServingState = ServingState.SHUTDOWN;
+                  future.set(throwable);
```

Comment:
It seems there still could be a race. This line can happen when `future.set(null)` is called, but changing the state to `STARTED` is not yet called. In this case `currentServingState = ServingState.SHUTDOWN` is called and then immediately after that the state is changed to `STARTED`.